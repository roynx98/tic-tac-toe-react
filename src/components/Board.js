import React from 'react';

import { Square } from './Square';

export class Board extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      squares: Array(9).fill(null),
      isXNext: true
    };
  }

  handleClick(i) {
    // Make a copy of the array
    const squares = this.state.squares.slice();

    squares[i] = this.state.isXNext ? 'X' : 'O';
    this.setState({
      squares,
      isXNext: !this.state.isXNext
    });
  }

  renderSquare(i) {
    return (
      <Square 
        value={this.state.squares[i]} 
        onClick={() => {this.handleClick(i)}}
        />
    );
  }

  render() {
    const status = `Next player: ${this.state.isXNext ? 'X' : 'O'}`;

    return (
      <div>
        <div className="status">{status}</div>
        <div className="board-row">
          {this.renderSquare(0)}
          {this.renderSquare(1)}
          {this.renderSquare(2)}
        </div>
        <div className="board-row">
          {this.renderSquare(3)}
          {this.renderSquare(4)}
          {this.renderSquare(5)}
        </div>
        <div className="board-row">
          {this.renderSquare(6)}
          {this.renderSquare(7)}
          {this.renderSquare(8)}
        </div>
      </div>
    );
  }
}