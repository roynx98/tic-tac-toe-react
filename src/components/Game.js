import { Board } from './Board';
import React from 'react';
import { calculateWinner } from '../helpers/calculateWinner';

export class Game extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      history: [{
        squares: Array(9).fill(null)
      }],
      isXNext: false
    };
  }

  handleClick(i) {
    const history = this.state.history;
    const current = history[history.length - 1];

    if (current.squares[i] ||Â calculateWinner(current.squares)) {
      return;
    }

    const newHistory = this.state.history.slice();
    const newSquares = current.squares.slice();

    newSquares[i] = this.state.isXNext ? 'X' : 'O';

    this.setState({
      history: newHistory.concat({ squares: newSquares }),
      isXNext: !this.state.isXNext
    });
  }

  render() {
    const history = this.state.history;
    const current = history[history.length - 1];
    const winner = calculateWinner(current.squares);

    let status = `Next player: ${this.state.isXNext ? 'X' : 'O'}`;
    if (winner) {
      status = `Winner: ${winner}`;
    }

    return (
      <>
        <div className="game">
          <div className="game-board">
            <Board 
              squares={current.squares}
              onClick={i => this.handleClick(i)} />
          </div>
          <div className="game-info">
            <div>{status}</div>
            <ol>{/* TODO */}</ol>
          </div>
        </div>
      </>
    );
  }

}
