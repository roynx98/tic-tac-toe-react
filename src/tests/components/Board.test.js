import React from 'react';
import { shallow } from 'enzyme';
import { Board } from '../../components/Board';
import sinon, { spy } from 'sinon';

describe('Board component', () => {
  const squares = [
    'X', null, 'O',
    null, 'O', null,
    null, null, 'X',
  ];

  test('should match snapshot', () => {
    const wrapper = shallow(<Board squares={squares} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should match squares', () => {
    const wrapper = shallow(<Board squares={squares} />);
    // First row
    expect(wrapper.find('.board-row').at(0).childAt(0).prop('value')).toBe(squares[0]);
    expect(wrapper.find('.board-row').at(0).childAt(1).prop('value')).toBe(squares[1]);
    expect(wrapper.find('.board-row').at(0).childAt(2).prop('value')).toBe(squares[2]);

    // Second row
    expect(wrapper.find('.board-row').at(1).childAt(0).prop('value')).toBe(squares[3]);
    expect(wrapper.find('.board-row').at(1).childAt(1).prop('value')).toBe(squares[4]);
    expect(wrapper.find('.board-row').at(1).childAt(2).prop('value')).toBe(squares[5]);

    // Third row
    expect(wrapper.find('.board-row').at(2).childAt(0).prop('value')).toBe(squares[6]);
    expect(wrapper.find('.board-row').at(2).childAt(1).prop('value')).toBe(squares[7]);
    expect(wrapper.find('.board-row').at(2).childAt(2).prop('value')).toBe(squares[8]);
  });

  test('should call a function on click a cell', () => {
    const onClickSpy = sinon.spy();
    const wrapper = shallow(<Board squares={squares} onClick={onClickSpy}/>);
    wrapper.find('.board-row').at(1).childAt(0).simulate('click');
    expect(onClickSpy.calledOnceWith(3)).toBe(true);
  });

});