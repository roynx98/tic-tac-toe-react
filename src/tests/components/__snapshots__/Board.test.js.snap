// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component should match snapshot 1`] = `
<div>
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value="X"
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value="O"
    />
  </div>
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value="O"
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </div>
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value="X"
    />
  </div>
</div>
`;
